### PR TITLE
Add CABLE model driver

### LINKED ISSUE #313: Add support for CABLE
Look into adding a driver for CABLE.

Part of the motivation was to help cleaning up the CABLE benchmarking suite

https://github.com/ccarouge/cbl_bench_carbon

ping @ccarouge

### PR DESCRIPTION
**Pull request by aidanheerdegen**
Add model driver for CABLE model

http://climate-cms.wikis.unsw.edu.au/CABLE

Closes #313

### CODE REVIEW

**Comment by aidanheerdegen** [comment]
@ccarouge I have implemented a basic templating scheme for the `cable.nml` file. 

There is an example of it's use in the `template` branch at this repo:

https://github.com/aidanheerdegen/cable_test/tree/template

The basic idea is to use `$year` in the `cable.nml` file, and read the value of `$year` from the previous restart, e.g. `archive/restart007/cable.res.yaml`, which for this example looks like so:

```yaml
year: 1904
```
It's still a bit rough. There is no way to start this process without a `cable.res.yaml` file existing, as I wasn't sure how to get the value of year otherwise. 

Just looked, and I can see it is available in the `restart.nc` file, so that would be the next step, interrogate that file if there is no `cable.res.yaml` file.

**Comment by aidanheerdegen** [comment]
I got notice that a CircleCI build failed. This is out of date because @dougiesquire has updated the CI to use GitHub Workflows. 

I/we should rebase this and force push to the repo to get it up to date with those changes. Note that there is a conflict in `payu/models/__init__.py` which needs to be manually resolved.

Happy to go through the process next week if that is helpful.

**Comment by SeanBryan51** [comment]
In [this payu experiment](https://github.com/CABLE-LSM/cable_example), we only have met forcing files from 1903 to 1909 - causing the model to crash when the year value exceeds 1909. Should we include logic to repeat the forcing once the model exceeds the duration of the forcing?

To implement this, I'm thinking we:
1. Replace `$year` with `$forcing_year` in the templating scheme to distinguish the 'model' year from the 'forcing' year.
2. Add an optional parameter `forcing_year` in cable.res.yaml which specifies the forcing year to use. If the forcing year is not specified, then the CABLE driver will use `year` as the forcing year.
3. The CABLE driver will somehow need to know the 'min' and 'max' forcing year to cycle over. To do this, we could add an optional config file: cable.cycle.yaml (add to `self.optional_config_files` attribute) which specifies the min/max forcing year.

@aidanheerdegen do you see any issues with this?

**Comment by aidanheerdegen** [comment]
@SeanBryan51 and I had a IRL chat. I believe I suggested a `setup` script could be used to update links to forcing files to make them cycle correctly.

Is that still an workable option?

**Comment by SeanBryan51** [comment]
I think I'm happy with the implementation of the driver. @aidanheerdegen or @jo-basevi are you happy to give this a review? Whenever you guys have the time.

**Comment by aidanheerdegen** [comment]
Happy to review .. so my first suggestion is to add some tests for your driver. Specifically the forcing logic, but if there are any other you can think of that might be useful add them to. There are some tests under `payu/test/models/` so take a look for some prior art if that helps.

No-one wants to hear "write tests", but you know you should :) .. particularly the guy that just gave a skillshare on `pytest`!

;)

@jo-basevi has written some nice ones recently using parameterised decorators if you want inspiration.

**Comment by SeanBryan51** [comment]
@aidanheerdegen I have added some tests for the forcing logic. I've tried to make the tests light weight by creating a separate helper function (the private methods in the `Cable` class had too much global state to write clean tests).

File: payu/models/cable.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import f90nml
+import yaml
+
+# Local
+from payu.fsops import mkdir_p
+from payu.models.model import Model
+
+
+def _get_forcing_path(variable, year, input_dir, offset=None, repeat=None):
+    """Return the met forcing file path for a given variable and year."""
```
Comment (aidanheerdegen) [inline payu/models/cable.py:25]: Some description of the arguments would be helpful. I see below that `offset` and `repeat` are 2-member tuples, but it isn't really obvious what they mean.

**Review by aidanheerdegen — COMMENTED** [review]
I didn't make any comments on the tests because I wasn't sure how to interpret the inputs.

File: payu/models/cable.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        super(Cable, self).setup()
+
+        self.cable_nml = f90nml.read(self.cable_nml_path)
+        if self.prior_restart_path:
+            with open(self.restart_calendar_path, 'r') as restart_file:
+                self.restart_info = yaml.safe_load(restart_file)
+        else:
+            self.restart_info = {'year': self.cable_nml['cable']['ncciy']}
+
+        year = self.cable_nml['cable']['ncciy'] = self.restart_info['year']
```
[the remark this answers is not in the document]
Comment (SeanBryan51) [reply payu/models/cable.py:93]: Your logic looks almost right, I think your else condition leaves `self.restart_info` uninitialised.

I can see how it looks confusing as we are assigning values to three things: `self.restart_info`, `self.cable_nml['cable']['ncciy']` and `year`. Is the following more readable?

```suggestion
        year = self.cable_nml['cable']['ncciy']
        self.restart_info = {'year': year}
        if self.prior_restart_path:
           with open(self.restart_calendar_path, 'r') as restart_file:
                self.restart_info = yaml.safe_load(restart_file)
            year = self.cable_nml['cable']['ncciy'] = self.restart_info['year']
```

File: payu/models/cable.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            os.path.join(self.work_path, self.cable_nml_fname),
+            force=True
+        )
+
+    def archive(self, **kwargs):
+
+        # Save model time to restart next run
+        with open(os.path.join(self.restart_path,
+                  self.restart_calendar_file), 'w') as restart_file:
+            restart = {'year': self.restart_info['year'] + 1}
```
[the remark this answers is not in the document]
Comment (SeanBryan51) [reply payu/models/cable.py:124]: I took a look at Jurgen's TRENDY experiments. The experiments seem to use per-year forcing so I guess I guessed wrong. Also, I realised Jurgen's setup uses a different version of CABLE which has other ways of reading in the met forcing which will probably complicate things. The driver currently supports only CABLE v3.

So far I haven't ported any CABLE runs which require the model to be run multiple years at a time so perhaps it is not a common use case. @ccarouge do you have any comments on this?

[the remark this answers is not in the document]
Comment (ccarouge) [reply payu/models/cable.py:155]: The whole way to read in the met forcings in CABLE needs to be redesigned. But this isn't going to happen now unless it throws a spanner in CABLE4 development. So the question is what do we do in the meantime.

I wouldn't look at multiple years now. Most (all?) of the setups I've seen are using yearly runs with an external loop for managing year changes. And that's the way it needs to run to deal with land-cover change I believe.

The fact the driver might not cover the TRENDY branch is more annoying but we can probably put changing the met forcing code in TRENDY branch at the top of the list (and test without payu and benchcab).

[the remark this answers is not in the document]
Comment (aidanheerdegen) [reply payu/models/cable.py:155]: I'll leave @SeanBryan51 and @ccarouge to resolve this, perhaps by adding an issue to the CABLE4 development?

**Comment by aidanheerdegen** [comment]
@SeanBryan51 are you ready to merge this? Do you need another review?

File: payu/models/cable.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SeanBryan51) [reply payu/models/cable.py:155]: I've marked this as resolved for now. If there is a need for running multiple years in the future then we can address the issue separately.

**Comment by SeanBryan51** [comment]
@aidanheerdegen after I clean up the commit history this should be all good to go in.

**Comment by SeanBryan51** [comment]
Just doing some final checks to make sure things are working.

**Comment by SeanBryan51** [comment]
Ready when you are @aidanheerdegen

**Review by SeanBryan51 — APPROVED** [review]
Approving on behalf of @aidanheerdegen

**Pull request merged**